test/opf/sdpwrm.jl

```diff
@@ -0,0 +1,222 @@
+    b_fr = [ref[:branch][e]["b_fr"] for e in 1:E]
+    b_to = [ref[:branch][e]["b_to"] for e in 1:E]
+    δθmin = [ref[:branch][e]["angmin"] for e in 1:E]
+    δθmax = [ref[:branch][e]["angmax"] for e in 1:E]
+
+    # Check dual feasibility for select buses and constraints
+    λp  = [res["solution"]["bus"]["$i"]["lam_kirchhoff_active"] for i in 1:N]
+    λq  = [res["solution"]["bus"]["$i"]["lam_kirchhoff_reactive"] for i in 1:N]
+    λpf = [res["solution"]["branch"]["$e"]["lam_ohm_active_fr"] for e in 1:E]
+    λqf = [res["solution"]["branch"]["$e"]["lam_ohm_reactive_fr"] for e in 1:E]
+    λpt = [res["solution"]["branch"]["$e"]["lam_ohm_active_to"] for e in 1:E]
+    λqt = [res["solution"]["branch"]["$e"]["lam_ohm_reactive_to"] for e in 1:E]
+
+    S = res["solution"]["S"]
+
+    μθ_lb = [res["solution"]["branch"]["$e"]["mu_va_diff_lb"] for e in 1:E]
+    μθ_ub = [-res["solution"]["branch"]["$e"]["mu_va_diff_ub"] for e in 1:E]
+
+    μ_w = [
+        res["solution"]["bus"]["$i"]["mu_w_lb"] + res["solution"]["bus"]["$i"]["mu_w_ub"]
+        for i in 1:N
+    ]
+
+    # Check dual constraint corresponding to `wr` variables
+    AR = sum(
+        (-gs[i] * λp[i] + bs[i] * λq[i]) * _get_sym(i, i, 1, N)
+        for i in 1:N;
+        init=0
+    )
+    + sum(
+        λpf[e] * (
+            (g[e]+g_fr[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["f_bus"], 1, N)
+            + (-g[e]*tr[e]+b[e]*ti[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["t_bus"], 1/2, N)
+        )
+        + λpt[e] * (
+            (g[e]+g_to[e]) * _get_sym(branch["t_bus"], branch["t_bus"], 1, N)
+            + (-g[e]*tr[e]-b[e]*ti[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["t_bus"], 1/2, N)
+        )
+        + λqf[e] * (
+            -(b[e]+b_fr[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["f_bus"], 1, N)
+            - (-b[e]*tr[e]-g[e]*ti[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["t_bus"], 1/2, N)
+        )
+        + λqt[e] * (
+            -(b[e]+b_to[e]) * _get_sym(branch["t_bus"], branch["t_bus"], 1, N)
+            - (-b[e]*tr[e]+g[e]*ti[e]) / ttm[e] * _get_sym(branch["f_bus"], branch["t_bus"], 1/2, N)
+        )
+        + (-tan(δθmin[e]) * μθ_lb[e] + tan(δθmax[e]) * μθ_ub[e]) * _get_sym(branch["f_bus"], branch["t_bus"], 1/2, N)
+        + μ_w[i] * _get_sym(branch["f_bus"], branch["f_bus"], 1, N)
+        for (e, branch) in ref[:branch];
+        init=0
+    )
+    @test norm(AR + S[1:N, 1:N] + S[(N+1):(2*N), (N+1):(2*N)], Inf) <= atol
+
+    # Check dual constraint corresponding to `wi` variables
+    AI = sum(
+        _get_skew_sym(branch["f_bus"], branch["t_bus"], 1/2, N) * (
+            λpf[e] * (-b[e]*tr[e]-g[e]*ti[e]) / ttm[e]
+            - λpt[e] * (-b[e]*tr[e]+g[e]*ti[e]) / ttm[e]
+            + λqf[e] * (-g[e]*tr[e]+b[e]*ti[e]) / ttm[e]
+            - λqt[e] * (-g[e]*tr[e]-b[e]*ti[e]) / ttm[e]
+            + μθ_lb[e]
+            - μθ_ub[e]
+        )
+        for (e, branch) in ref[:branch];
+        init=0
+    )
```

Comment:
Similar comment as before regarding the performance of summing large numbers of single-entry sparse matrices